### ISSUE TITLE
Gradient operations (zero_grad and gradient accumulations) as graphs

### ISSUE THREAD

**Issue by kbadz-pl** [pytorch/pytorch#103840]
### 🚀 The feature, motivation and pitch

Hi,

We already can see in PT2.0 that operations which are not part of main FWD/BWD/OPT blocks - especially **zero_grad** and **gradient accumulation** - can become serious host bottleneck due to being scaled with topology size and executed eagerly.

Are there any plans to incorporate these operations into the graphs in PT2.1 or later? Thanks.

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7 @msaroufim @wconstab @bdhirsh @anijain2305

**Comment by jansel** [comment]
Yes that is correct.

**Comment by kbadz-pl** [comment]
That's great to hear. Will this feature make it into PT2.1 release?

**Comment by albanD** [comment]
For the zero_grad side, do you use `set_to_none=True` already? That can reduce the compute significantly in some cases.

**Comment by jansel** [comment]
> That's great to hear. Will this feature make it into PT2.1 release?

We are not pushing for 2.1 on this feature.  The plan is to ship it when it is ready.

**Comment by jansel** [comment]
Related PR is here: https://github.com/pytorch/pytorch/pull/103822

**Comment by kbadz-pl** [comment]
> For the zero_grad side, do you use `set_to_none=True` already? That can reduce the compute significantly in some cases.

@albanD  Yes, but these zeroing operations still seem to occur right before the compiled backward graph:
![image](https://github.com/pytorch/pytorch/assets/122799060/1840978e-57b7-4c88-a63f-5c75c5d66010)

Is this expected (do you know where it comes from?) ? And would that be addressed by above PR?

**Comment by albanD** [comment]
These are calls to `aten:zeros` not `aten::zero_`. So I don't think they are the same thing. I guess this is the part that creates the output of the compute.

**Comment by albanD** [comment]
Given the original ask above, I guess these are all the parameters? So they are needed?

**Comment by Chillee** [comment]
Filed an issue for the "zeroing out non-differentiable outputs" part: https://github.com/pytorch/pytorch/issues/104272

**Comment by kbadz-pl** [comment]
> Related PR is here: #103822

@jansel I see that this PR was merged. Does it mean I should no longer see the grad accumulations in the traces on latest master?

**Comment by ezyang** [comment]
No, compiled backward is not on by default. Try turning it on for your use case and tell us if it works or not.

**Comment by kbadz-pl** [comment]
@ezyang got it, how can I enable this feature?

**Comment by ezyang** [comment]
Check out https://github.com/pytorch/pytorch/pull/103822

**Comment by bdhirsh** [comment]
I'm going to tentatively close this issue, since you can now capture gradient accumulation and zero_grad if you use compiled autograd. This is an opt-in API though, so you won't get this behavior "by default" (I don't think we have plans to change this)
